posthog/models/exchange_rate/sql.py

```diff
@@ -0,0 +1,166 @@
+                            pass
+
+                rates_dict[date_key] = rates
+            except ValueError:
+                # Skip rows with invalid dates
+                continue
+
+    # Fill in missing dates
+    start_date = datetime.datetime(2000, 1, 1)
+    end_date = datetime.datetime(2024, 12, 31)
+    current_date = start_date
+
+    while current_date <= end_date:
+        date_str = current_date.strftime("%Y-%m-%d")
+
+        if date_str not in rates_dict:
+            rates_dict[date_str] = {}
+
+        current_date += datetime.timedelta(days=1)
+
+    return rates_dict
+
+
+# Yield from HISTORICAL_EXCHANGE_RATE_DICTIONARY()
+# a tuple in the form of (date, currency, rate)
+def HISTORICAL_EXCHANGE_RATE_TUPLES():
+    rates_dict = HISTORICAL_EXCHANGE_RATE_DICTIONARY()
+    for date, rates in rates_dict.items():
+        for currency, rate in rates.items():
+            yield (date, currency, rate)
+
+
+EXCHANGE_RATE_TABLE_NAME = "exchange_rate"
+EXCHANGE_RATE_DICTIONARY_NAME = "exchange_rate_dict"
+
+# `version` is used to ensure the latest version is kept, see `MergeTreeEngine`
+EXCHANGE_RATE_TABLE_SQL = (
+    lambda on_cluster=True: """
+CREATE TABLE IF NOT EXISTS {table_name} {on_cluster_clause} (
+    date Date,
+    currency String,
+    rate Decimal64(4),
+    version UInt32 DEFAULT toUnixTimestamp(now())
+) ENGINE = {engine}
+ORDER BY (date, currency);
+""".format(
+        table_name=EXCHANGE_RATE_TABLE_NAME,
+        engine=MergeTreeEngine(
+            "exchange_rate", replication_scheme=ReplicationScheme.REPLICATED, replacing_col="version"
+        ),
+        on_cluster_clause=ON_CLUSTER_CLAUSE(on_cluster),
+    )
+)
+
+
+DROP_EXCHANGE_RATE_TABLE_SQL = f"DROP TABLE IF EXISTS {EXCHANGE_RATE_TABLE_NAME} ON CLUSTER '{CLICKHOUSE_CLUSTER}'"
+
+
+TRUNCATE_EXCHANGE_RATE_TABLE_SQL = (
+    f"TRUNCATE TABLE IF EXISTS {EXCHANGE_RATE_TABLE_NAME} ON CLUSTER '{CLICKHOUSE_CLUSTER}'"
+)
+
+
+def EXCHANGE_RATE_DATA_BACKFILL_SQL(exchange_rates=None):
+    if exchange_rates is None:
+        exchange_rates = HISTORICAL_EXCHANGE_RATE_TUPLES()
+
+    return f"""
+INSERT INTO exchange_rate (date, currency, rate) VALUES
+{
+''',
+'''.join(f"(toDate('{date}'), '{currency}', {rate})" for date, currency, rate in exchange_rates)},
+;
+"""
+
+
+# Use COMPLEX_KEY_HASHED, as we have a composite key
+# Also, note the `anyLast` function, which is used to get the latest rate for a given date and currency
+# given that we might have more than one while the merges haven't finished yet
+EXCHANGE_RATE_DICTIONARY_SQL = (
+    lambda on_cluster=True: f"""
+CREATE DICTIONARY IF NOT EXISTS {EXCHANGE_RATE_DICTIONARY_NAME} {ON_CLUSTER_CLAUSE(on_cluster)} (
+    date Date,
+    currency String,
+    rate Decimal64(4)
+)
+PRIMARY KEY (date, currency)
+SOURCE(CLICKHOUSE(TABLE '(SELECT date, currency, anyLast(rate) AS rate FROM {EXCHANGE_RATE_TABLE_NAME} GROUP BY date, currency)' PASSWORD '{CLICKHOUSE_PASSWORD}'))
```

Comment:
why passing a password here is necessary?

Reply:
ClickHouse requires it, otherwise it will fail, even though it's querying itself... 🤷

Reply:
Also I think you want to use QUERY instead of table here. And I'd say that parenthesis are not needed:

```suggestion
SOURCE(CLICKHOUSE(QUERY 'SELECT date, currency, anyLast(rate) AS rate FROM {EXCHANGE_RATE_TABLE_NAME} GROUP BY date, currency' PASSWORD '{CLICKHOUSE_PASSWORD}'))
```

Reply:
Oh, you're right, changing to QUERY